core/templates/services/playthrough-issues.service.spec.ts

```diff
@@ -0,0 +1,241 @@
+      'PlaythroughIssuesBackendApiService');
+    ImprovementModalService = $injector.get(
+      'ImprovementModalService');
+    $q = $injector.get('$q');
+    $rootScope = $injector.get('$rootScope');
+
+    PlaythroughIssuesService.initSession(explorationId, explorationVersion);
+  }));
+
+  it('should get issues from backend', function() {
+    var backendCallSpy = spyOn(
+      playthroughIssuesBackendApiService, 'fetchIssues').and.callThrough();
+
+    PlaythroughIssuesService.getIssues().then(function(issues) {
+      expect(backendCallSpy).toHaveBeenCalled();
+      expect(issues).toEqual(
+        backendIssues.map(
+          PlaythroughIssueObjectFactory.createFromBackendDict));
+    });
+  });
+
+  it('should get playthrough from backend', function() {
+    var backendCallSpy = spyOn(
+      playthroughIssuesBackendApiService, 'fetchPlaythrough').and
+      .callThrough();
+    var playthroughId = 'exp_id1';
+
+    PlaythroughIssuesService.getPlaythrough(playthroughId)
+      .then(function(playthrough) {
+        expect(backendCallSpy).toHaveBeenCalled();
+        expect(playthrough).toEqual(
+          PlaythroughObjectFactory.createFromBackendDict(backendPlaythrough));
+      });
+  });
+
+  it('should render issue statement when its type is' +
+  ' MultipleIncorrectSubmissions', function() {
+    var copiedBackendIssue = angular.copy(backendIssues[0]);
+    var issue = PlaythroughIssueObjectFactory.createFromBackendDict(
+      copiedBackendIssue);
+    var stateName = issue.issueCustomizationArgs.state_name.value;
+
+    expect(PlaythroughIssuesService.renderIssueStatement(issue))
+      .toBe('Several learners submitted answers to card "' + stateName +
+      '" several times, then gave up and quit.');
+  });
+
+  it('should render issue statement when its type is EarlyQuit', function() {
+    var copiedBackendIssue = angular.copy(backendIssues[0]);
+    copiedBackendIssue.issue_type = 'EarlyQuit';
+    var issue = PlaythroughIssueObjectFactory.createFromBackendDict(
+      copiedBackendIssue);
+
+    expect(PlaythroughIssuesService.renderIssueStatement(issue))
+      .toBe('Several learners exited the exploration in less than a minute.');
+  });
+
+  it('should render issue statement when its type is CyclicStateTransitions',
+    function() {
+      var copiedBackendIssue = angular.copy(backendIssues[0]);
+      copiedBackendIssue.issue_type = 'CyclicStateTransitions';
+      var issue = PlaythroughIssueObjectFactory.createFromBackendDict(
+        copiedBackendIssue);
+      var stateName = issue.issueCustomizationArgs.state_names.value[0];
+
+      expect(PlaythroughIssuesService.renderIssueStatement(issue))
+        .toBe('Several learners ended up in a cyclic loop revisiting card "' +
+        stateName + '" many times.');
+    });
+
+  it('should render issue suggestion when its type is' +
+    ' MultipleIncorrectSubmissions', function() {
+    var copiedBackendIssue = angular.copy(backendIssues[0]);
```

Comment:
Instead of making copies, please use Jasmine's user context features by storing it in `this.backendIssues`. Jasmine will reset changes to any and all objects assigned to `this` between each test case: [source](https://gist.github.com/traviskaufman/11131303).

Reply:
I'm fine with not doing this here to unblock you, but please file it as an issue so we can continue discussion about it.

Reply:
Please file this as an issue and then I'll LGTM, thanks!

Reply:
Done!